cirq/value/states.py

```diff
@@ -0,0 +1,365 @@
+
+    def __getitem__(self, qubit):
+        """Return the NamedState at the given qubit."""
+        return self.states[qubit]
+
+    def __iter__(self):
+        yield from self.states.items()
+
+    def __len__(self):
+        return len(self.states)
+
+    def __eq__(self, other):
+        if not isinstance(other, TensorProductState):
+            return False
+
+        return self.states == other.states
+
+    def __hash__(self):
+        return hash(tuple(self.states.items()))
+
+    def _json_dict_(self):
+        return {
+            'cirq_type': self.__class__.__name__,
+            'states': list(self.states.items()),
+        }
+
+    @classmethod
+    def _from_json_dict_(cls, states, **kwargs):
+        return cls(states=dict(states))
+
+    def state_vector(self, qubit_order: 'cirq.QubitOrder' = None):
+        """The state-vector representation of this state."""
+        from cirq import ops
+        if qubit_order is None:
+            qubit_order = ops.QubitOrder.DEFAULT
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(self.qubits)
+
+        mat = 1.0 + 0.j
+        for qubit in qubits:
+            oneq_state = self[qubit]
+            state_vector = oneq_state.state_vector()
+            mat = np.kron(mat, state_vector)
+
+        return mat
+
+    def projector(self, qubit_order: 'cirq.QubitOrder' = None):
+        """The projector associated with this state expressed as a matrix.
+
+        This is |s><s| where |s> is this state.
+        """
+        from cirq import ops
+        if qubit_order is None:
+            qubit_order = ops.QubitOrder.DEFAULT
+        qubit_order = ops.QubitOrder.as_qubit_order(qubit_order)
+        qubits = qubit_order.order_for(self.qubits)
+
+        mat = 1.0 + 0.j
+        for qubit in qubits:
+            oneq_state = self[qubit]
+            oneq_proj = oneq_state.projector()
+            mat = np.kron(mat, oneq_proj)
+        return mat
+
+
+class _PlusX(_Named1QState):
+
+    def __str__(self):
+        return '+X'
+
+    def __repr__(self):
+        return 'cirq.PLUS_X'
+
+    def state_vector(self):
+        return np.array([1, 1]) / np.sqrt(2)
+
+    def stabilized_by(self):
+        from cirq import ops
+        return 1, ops.X
+
+
+class _MinusX(_Named1QState):
+    def __str__(self):
+        return '-X'
+
+    def __repr__(self):
+        return 'cirq.MINUS_X'
+
+    def state_vector(self):
+        return np.array([1, -1]) / np.sqrt(2)
+
+    def stabilized_by(self):
+        from cirq import ops
+        return -1, ops.X
+
+
+class _PlusY(_Named1QState):
+    def __str__(self):
+        return '+Y'
+
+    def __repr__(self):
+        return 'cirq.PLUS_Y'
+
+    def state_vector(self):
+        return np.array([1, 1j]) / np.sqrt(2)
+
+    def stabilized_by(self):
+        from cirq import ops
+        return 1, ops.Y
+
+
+class _MinusY(_Named1QState):
+
+    def __str__(self):
+        return '-Y'
```

Comment:
These methods could be attributes of a concrete base class, instead of implementations of abstract methods of an abstract base class.

Reply:
This is modeled after how e.g. `cirq.X` and friends work. The alternative (as you suggest) would be a concrete base, e.g.

```python
KET_PLUS = _Named1QState(
    name='KET_PLUS',
    str='+X',
    state_vector = np.array([1,1])/np.sqrt(2),
    stabilized_by=(+1, ops.X)
)
```

big problem: stabilized by needs to import `cirq.ops` which would cause a circular import because of `value_equality`.

Small problem: we'd be initializing the arrays at import time. We don't do this anywhere else

Other consideration: It would be marginally easier with this scheme to input arbitrary or nonsensical values for `name`, `str`, etc. whereas it makes more sense to consider this as "enum-like"